NIFI-9822 - ConsumeKafkaRecord allows writing out Kafka record key

greyp9 opened the pull request:
# Summary

[NIFI-9822](https://issues.apache.org/jira/browse/NIFI-9822)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 8
  - [X] JDK 11
  - [X] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [X] Documentation formatting appears as expected in rendered files

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/ConsumerLease.java:
```diff
@@ -653,6 +672,81 @@ private void writeRecordData(final ProcessSession session, final List<ConsumerRe
+    }
+
+    private Tuple<RecordField, Object> toWrapperRecordMetadata(final ConsumerRecord<byte[], byte[]> consumerRecord) {
+        final RecordField fieldTopic = new RecordField("topic", RecordFieldType.STRING.getDataType());
+        final RecordField fieldPartition = new RecordField("partition", RecordFieldType.INT.getDataType());
+        final RecordField fieldOffset = new RecordField("offset", RecordFieldType.LONG.getDataType());
+        final RecordField fieldTimestamp = new RecordField("timestamp", RecordFieldType.TIMESTAMP.getDataType());
+        final RecordSchema schema = new SimpleRecordSchema(Arrays.asList(
+                fieldTopic, fieldPartition, fieldOffset, fieldTimestamp));
```
This schema and all of its field can be statically defined once, rather than instantiating all of this for every invocation.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/KafkaProcessorUtils.java:
```diff
@@ -110,6 +110,12 @@ public final class KafkaProcessorUtils {
             "For dataflows where ordering of FlowFiles is important, this strategy can be used along with ensuring that the each processor in the dataflow uses only a single Concurrent Task.");
 
+    static final AllowableValue WRITE_VALUE_ONLY = new AllowableValue("write-value-only", "Write Value Only", "Write only the Kafka Record value.");
+    static final AllowableValue USE_WRAPPER = new AllowableValue("use-wrapper", "Use Wrapper", "Write the Kafka Record key, value, headers, and metadata.");
```
Perhaps we should mention in the description explicitly that these fields will be wrapped in a wrapper element, and to see Processor Usage for more information.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/PublishKafkaRecord_2_6.java:
```diff
@@ -271,7 +273,21 @@ public class PublishKafkaRecord_2_6 extends AbstractProcessor implements Verifia
+    static final PropertyDescriptor PUBLISH_STRATEGY = new PropertyDescriptor.Builder()
+            .name("publish-strategy")
+            .displayName("Publish Strategy")
+            .description("The format used to publish the outgoing FlowFile record to Kafka.")
```
I would avoid the term "outgoing FlowFile" because that term is generally used to refer to a FlowFile that is being transferred to an outbound relationship. Here, it's not an outgoing FlowFile that we're publishing to Kafka. It's an incoming FlowFile.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/PublishKafkaRecord_2_6.java:
```diff
@@ -271,7 +273,21 @@ public class PublishKafkaRecord_2_6 extends AbstractProcessor implements Verifia
+            .description("The format used to publish the outgoing FlowFile record to Kafka.")
+            .required(true)
+            .defaultValue(WRITE_VALUE_ONLY.getValue())
+            .allowableValues(WRITE_VALUE_ONLY, USE_WRAPPER)
```
I'm not sure that "Write Value Only" is applicable here. WIth this strategy, we can still send message headers using attributes, and a message key by specifying a field name.

greyp9 replied:
I understand you're suggesting a different name.  What would that be?

option:
- WRITE_VALUE, WRITE_WRAPPER

Or are you saying there shouldn't be an option?

markap14 replied:
Yes, sorry, I was suggesting renaming it. Perhaps something like "Use Content as Record Value" or "FLowFile Content as Record Value" or something like that? Was just trying to convey that we aren't *only* writing the record value, a key and headers may also be written.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/PublishKafkaRecord_2_6.java:
```diff
@@ -271,7 +273,21 @@ public class PublishKafkaRecord_2_6 extends AbstractProcessor implements Verifia
         .required(false)
         .build();
-
+    static final PropertyDescriptor PUBLISH_STRATEGY = new PropertyDescriptor.Builder()
```
With the addition of this property, I think some of the existing properties can also be hidden if using the wrapper. Specifically, just like in the consumer, we should make MESSAGE_KEY_FIELD and ATTRIBUTE_NAME_REGEX dependent properties because it doesn't make sense to specify those if using the Wrapper.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/PublisherLease.java:
```diff
@@ -193,10 +196,24 @@ void publish(final FlowFile flowFile, final RecordSet recordSet, final RecordSet
+                            .map(s -> s.getBytes(StandardCharsets.UTF_8)).orElse(null);
+                } else {
+                    try (final ByteArrayOutputStream os = new ByteArrayOutputStream(1024)) {
+                        final Record keyRecord = Optional.ofNullable(record.getValue(messageKeyField))
+                                .filter(Record.class::isInstance).map(Record.class::cast)
+                                .orElseThrow(() ->  new IOException("The property 'Record Key Writer' is defined, but the record key is not a record"));
```
Not sure that it makes sense to throw an Exception in this case. If the key is a `String`, we should just convert it to a `byte[]` via `String.toByteArray(StandardCharsets.UTF_8)`. If the key is a `byte[]` already then just send it as-is. We should only use the Record Reader when necessary.
In short - we should be explicit in what we produce, and lenient in what we accept.

markap14 replied:
I think I would actually do something like:
```
final Object key;
final Object keyValue = record.getValue(messageKeyField);
if (keyValue == null) {
  key = keyValue;
} else if (keyValue instanceof byte[]) {
  key = (byte[]) keyValue;
} else if (keyValue instanceof Byte[]) {
  // This case exists because in our Record API we currently don't have a BYTES type, we use an Array of type Byte, which creates a Byte[] instead of a byte[]. We should address this in the future, but we should account for the log here.
  final Byte[] bytes = (Byte[]) keyValue;
  key = new byte[bytes.length];
  for (int i=0; i < bytes.length; i++) {
    key[i] = bytes[i];
  }
} else if (key instanceof Record) {
 ... // what you have below, to create record writer, write key, flush
} else {
  final String keyString = keyValue.toString();
  key = keyString.getBytes(StandardCharsets.UTF_8);
}
```
I.e., just look at the object that you have. Using the schema is also feasible. But it can get really complicated if the schema indicates that it's a CHOICE type because then you'll probably still end up having to do this type of logic.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/resources/docs/org.apache.nifi.processors.kafka.pubsub.ConsumeKafkaRecord_2_6/additionalDetails.html:
```diff
@@ -256,5 +256,69 @@ <h3>SASL_SSL</h3>
+            </ul>
+
+
+            <p>The record schema that is used when 'Use Wrapper' is active is as follows:</p>
+            <ul>
+                <li>key: one of RECORD, STRING, or BYTE_ARRAY (based on processor configuration property 'Key Format')
+                </li>
+                <li>value: RECORD</li>
+                <li>headers: MAP (STRING, STRING)</li>
+                <li>metadata: RECORD
+                    <ul>
+                        <li>topic: STRING</li>
+                        <li>partition: INTEGER</li>
+                        <li>offset: LONG</li>
+                        <li>timestamp: TIMESTAMP</li>
+                    </ul>
+                </li>
+            </ul>
```
Is OK to provide the schema this way, as it's fairly easy to understand. But we should also provide the schema as an Avro Schema, since that's what the user will likely need to configure

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/resources/docs/org.apache.nifi.processors.kafka.pubsub.ConsumeKafkaRecord_2_6/additionalDetails.html:
```diff
@@ -256,5 +256,69 @@ <h3>SASL_SSL</h3>
+                Wrapper' includes headers and keys in the FlowFile content, they are not also added to the FlowFile
+                attributes. These properties are available only when the FlowFile Output Strategy is set to 'Write
+                Value Only'.</p>
+        </div>
```
I'd recommend providing an example output that was written using the JsonRecordSetWriter. This will help to clarify exactly what would be produced by the processor.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/resources/docs/org.apache.nifi.processors.kafka.pubsub.PublishKafkaRecord_2_6/additionalDetails.html:
```diff
@@ -189,5 +189,30 @@ <h3>SASL_SSL</h3>
+                downstream data processing (including that available in ConsumeKafkaRecord_2_6). If not defined, the
+                default implementation serializes the record key as an unmodified byte array (unchanged from previous
+                versions of the processor).</p>
+        </div>
```
Would recommend showing the same example that I referred to in ConsumeKafkaRecord_2_6, where the payload is in JSON, and then showing exactly what would be sent to Kafka, what would be used as the key, and the headers. For the metadata, we have a field for the topic also. Would that be used here? Would any of the 'metadata' fields be used? Or would they all be ignored?

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/test/java/org/apache/nifi/processors/kafka/pubsub/TestConsumeKafkaMock.java:
```diff
@@ -0,0 +1,520 @@
+        final JsonNode nodeFlowFile = mapper.readTree(flowFile.getContent());
+        assertTrue(nodeFlowFile instanceof ArrayNode);
+        assertEquals(1, nodeFlowFile.size());
+        // extract the NiFi json object representation of Kafka input record
+        final JsonNode flowFileValue = nodeFlowFile.iterator().next();
+        // wrapper object contains "key", "value", "headers", "metadata"
+        assertEquals(4, flowFileValue.size());
+        final JsonNode nodeWrapperKey = flowFileValue.get("key");
+        final JsonNode nodeWrapperValue = flowFileValue.get("value");
+        final JsonNode nodeWrapperHeaders = flowFileValue.get("headers");
+        final JsonNode nodeWrapperMetadata = flowFileValue.get("metadata");
+        assertNotNull(nodeWrapperKey);
+        assertNotNull(nodeWrapperValue);
+        assertNotNull(nodeWrapperHeaders);
+        assertNotNull(nodeWrapperMetadata);
```
A lot of the assertions in this test appear to be very repetitive. Especially this section here. Perhaps it makes sense to pull this out into a separate method, `void assertWrapperElements(JsonNode wrapper)`?

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/test/java/org/apache/nifi/processors/kafka/pubsub/TestPublishKafkaMock.java:
```diff
@@ -0,0 +1,339 @@
+        runner.enqueue(flowFile);
+        runner.run(1);
+        // verify results
+        runner.assertTransferCount("success", 1);
+        runner.assertTransferCount("failure", 0);
```
This can be simplified as `runner.assertAllFlowFilesTransferred("success", 1);`. This ensures that 1 went to success and also that no FlowFiles went anywhere else. Not particularly important, just wanted to point it out.

greyp9 replied:
Nice!

greyp9 commented:
> Thanks for pushing this @greyp9 ! I think this gets us really far down field. I left some comments inline, mostly around terminology / naming conventions, and documentation. I also think the tests that were added are really helpful, but I feel like some of the code is super repetitive and some of the sets of assertions can be easily refactored into a method so that the code is a little cleaner & less verbose - unless I missed some discrepancy between very similar chunks of code that are actually different.

Just pushed another commit; it should address the issues you noted.

There were some efficiencies to be gained in the unit tests, but the variables needed to verify behavior made the suggested refactor difficult.  Hope the updates reach the needed bar.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/resources/docs/org.apache.nifi.processors.kafka.pubsub.ConsumeKafkaRecord_2_6/additionalDetails.html:
```diff
@@ -256,5 +256,133 @@ <h3>SASL_SSL</h3>
+    "name": "kafka:ConsumeRecord:metadata",
+    "namespace": "org.apache.nifi",
+    "fields": [{
+      "name": "key",
+      "type": ["bytes", "string", "record"]
+    }, {
+      "name": "topic",
+      "type": "string"
+    }, {
+      "name": "partition",
+      "type": "int"
+    }, {
+      "name": "offset",
+      "type": "long"
+    }, {
+      "name": "timestamp",
+      "type": "long",
+      "logicalType": "timestamp-millis"
+    }]
+  },
+  {
+    "type": "record",
+    "name": "kafka:ConsumeRecord:wrapper",
+    "namespace": "org.apache.nifi",
+    "fields": [{
+      "name": "key",
+      "type": ["bytes", "string", "record"]
+    }, {
+      "name": "value",
+      "type": "record"
+    }, {
+      "name": "headers",
+      "type": "map",
+      "values": "string"
+    }, {
+      "name": "metadata",
+      "type": "kafka:ConsumeRecord:metadata"
+    }]
+  }
+]
```
I think I would document the schema as:
```
{
  "type": "record",
  "name": "nifiRecord",
  "namespace": "org.apache.nifi",
  "fields": [{
      "name": "key",
      "type": [{
          <Schema is determined by the Key Record Reader, or will be "string" or "bytes", depending on the "Key Format" property>
        }, "null"]
    },
    {
      "name": "value",
      "type": [
        {
          <Schema is determined by the Value Record Reader, or will be "string" or "bytes", depending on the "Key Format" property>
        },
        "null"
      ]
    },
    {
      "name": "headers",
      "type": [
        { "type": "map", "values": "string", "default": {}},
        "null"]
    },
    {
      "name": "metadata",
      "type": [
        {
          "type": "record",
          "name": "metadataType",
          "fields": [
            { "name": "topic", "type": ["string", "null"] },
            { "name": "partition", "type": ["int", "null"] },
            { "name": "offset", "type": ["int", "null"] },
            { "name": "timestamp", "type": ["long", "null"] }
          ]
        },
        "null"
      ]
    }
  ]
}
```

greyp9 replied:
I had used this as a guide for expressing multiple schemas:
https://stackoverflow.com/a/40865366

Will replace with your suggestion; thanks!

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/PublishKafkaRecord_2_6.java:
```diff
@@ -271,7 +284,13 @@ public class PublishKafkaRecord_2_6 extends AbstractProcessor implements Verifia
-
+    static final PropertyDescriptor RECORD_KEY_WRITER = new PropertyDescriptor.Builder()
+            .name("record-key-writer")
+            .displayName("Record Key Writer")
```
Given that we're adding a Record Key Writer, we should probably change display name of "Record Writer" to "Record Value Writer" for clarity

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/PublishKafkaRecord_2_6.java:
```diff
@@ -312,6 +331,8 @@ public class PublishKafkaRecord_2_6 extends AbstractProcessor implements Verifia
         properties.add(KafkaProcessorUtils.TOKEN_AUTH);
         properties.add(KafkaProcessorUtils.SSL_CONTEXT_SERVICE);
         properties.add(MESSAGE_KEY_FIELD);
+        properties.add(PUBLISH_STRATEGY);
```
Is probably best to add Publish Strategy before Message Key Field, since Message Key Field depends on Publish Strategy. Otherwise, we have a situation where changing the value of Publish Strategy introduce a new property above it, which is kind of an odd user experience.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/ConsumerLease.java:
```diff
@@ -593,6 +609,9 @@ private void writeRecordData(final ProcessSession session, final List<ConsumerRe
                     try {
                         Record record;
                         while ((record = reader.nextRecord()) != null) {
```
There's a special case that we need to consider here.
It's possible for a Kafka record to have a null value. In that case, it will produce no NiFi Records using the Record Reader. However, if using the Output Strategy of Wrapper, we should still output the wrapper. It may have a Key or Headers that are still of value.
So here, we should keep track of the fact that our Record Reader has produced a Record. When we end this `while` loop, if no NiFi Record was produced, we should output a Record where the value is `null` but the Key and Headers are still populated.

greyp9 replied:
Rather than tracking state, would it be reasonable to predetermine the decision based on `consumerRecord.value()` a few lines above?

```
final byte[] recordBytes = consumerRecord.value() == null ? new byte[0] : consumerRecord.value();
```

(If it is possible to emit no record even with non-null data, I guess this wouldn't work.)

markap14 replied:
It's possible to have bytes that produce no record. For example, it could be JSON and contain `{}` or it could be a CSV message with just a header line and no actual data.

greyp9 replied:
It looks like it would be necessary to extract most of the logic inside the while loop into a helper method, as it would also be needed to process this null record.  Am I missing a simpler way to do this?

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/ConsumeKafkaRecord_2_6.java:
```diff
@@ -426,11 +466,13 @@ protected ConsumerPool createConsumerPool(final ProcessContext context, final Co
 
             return new ConsumerPool(maxLeases, readerFactory, writerFactory, props, topics, maxUncommittedTime, securityProtocol,
-                bootstrapServers, log, honorTransactions, charset, headerNamePattern, separateByKey, keyEncoding, partitionsToConsume, commitOffsets);
+                    bootstrapServers, log, honorTransactions, charset, headerNamePattern, separateByKey, keyEncoding, partitionsToConsume,
+                    commitOffsets, outputStrategy, keyFormat, keyReaderFactory);
```
While it makes sense to have `String outputStrategy` at this level, as it was obtained via Property value, it can only have 2 legitimate values. As-is, the ConsumerPool has to then compare the value using `KafkaProcessorUtils.OUTPUT_USE_WRAPPER.getValue().equals(consumeStrategy)`.
It would probably be much cleaner to instead have `enum` for the possible Output Strategies and pass that enum value into the ConsumerPool.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/PublishKafkaRecord_2_6.java:
```diff
@@ -239,6 +251,7 @@ public class PublishKafkaRecord_2_6 extends AbstractProcessor implements Verifia
             + "If not specified, no FlowFile attributes will be added as headers.")
         .addValidator(StandardValidators.REGULAR_EXPRESSION_VALIDATOR)
         .expressionLanguageSupported(NONE)
+        .dependsOn(PUBLISH_STRATEGY, PUBLISH_USE_VALUE)
```
If this property now depends on Publish Strategy, it should come after Publish Strategy in the Properties list (i.e., in `this.properties`)

greyp9 replied:
Do you think a PublishStrategy enum would be useful here (like OutputStrategy suggested above)?

greyp9 replied:
Oof.  That's an oversight.  The propagation from processor to pool to lease is needed, so I'll do enum here as well.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/resources/docs/org.apache.nifi.processors.kafka.pubsub.PublishKafkaRecord_2_6/additionalDetails.html:
```diff
@@ -189,5 +189,77 @@ <h3>SASL_SSL</h3>
+                default implementation serializes the record key as an unmodified byte array (unchanged from previous
+                versions of the processor).</p>
+
+<p>Here is an example of a Kafka Record that is emitted by <code>JsonRecordSetWriter</code> when strategy "Use Wrapper" is active:</p>
```
I think this section is inaccurate. This describes how the processor would behave if the "Publish Strategy" is set to "Use Content as Record Value", not "Use Wrapper".
I do think this is as good example, as it illustrates how a key can be specified, etc. But we should also have an example of how to use the Publish Strategy of "Use Wrapper". In this case, we wouldn't specify the Key Field, and the FlowFile content would be expected to have content that matches the wrapper's format. E.g.,
```
{
  "key": {"name":"Acme","number":"AC1234"},
  "value": {"address":"1234 First Street","zip":"12345","account":{"name":"Acme","number":"AC1234"}},
  "headers": {}
}
```

markap14 replied:
It also is important here to note that the `metadata` field is currently ignored, if it is.
However, I don't think we should actually ignore the field. Or, at least, we should optionally ignore it. It would make sense to use both the `topic` and the `partition` values of the metadata fields. This would allow us to send in data that goes to multiple topics, even as a single transaction, which would be a pretty nice feature.
It's okay to ignore it for now, as long as the documentation very clearly & explicitly states that we ignore that field.

greyp9 replied:
Updated wrapper record to include metadata fields topic and timestamp.  If I pulled partition up to the beginning of the method, that could be included as well, but I don't know whether that would affect the application of the partition strategy.

markap14 reviewed (changes requested):
Thanks for updating this @greyp9. I think we're getting close! I got a chance to sit down with it and try it out for a bit. I made several comments inline. Most of them are trivial usability updates (orders of properties, property display names, etc.) or documentation related. A few may take a bit more consideration, though.

markap14 commented:
Thanks for updating @greyp9 . Trying this out again.
I just tried sending the following content via PublishKafkaRecord_2_6:
```
{
  "key": {
    "type": "person"
  },
  "value": {
    "name": "Mark",
    "number": 49
  },
  "headers": {
     "headerA": "headerAValue"
  }
}
```
I set Publish Strategy to "Use Wrapper" and used a JsonTreeReader as the record reader.
But I encountered a NullPointerException:
```
2022-08-22 15:58:22,792 ERROR [Timer-Driven Process Thread-2] o.a.n.p.k.pubsub.PublishKafkaRecord_2_6 PublishKafkaRecord_2_6[id=c71d2b54-0182-1000-4979-53ab4cc8d555] Failed to send StandardFlowFileRecord[uuid=e8d03807-fa13-4620-be6a-72e014ac0838,claim=StandardContentClaim [resourceClaim=StandardResourceClaim[id=1661197835339-1, container=default, section=1], offset=3200, length=145],offset=0,name=e8d03807-fa13-4620-be6a-72e014ac0838,size=145] to Kafka
java.lang.NullPointerException: null
	at org.apache.nifi.processors.kafka.pubsub.PublisherLease.toWrapperRecord(PublisherLease.java:262)
	at org.apache.nifi.processors.kafka.pubsub.PublisherLease.publish(PublisherLease.java:210)
	at org.apache.nifi.processors.kafka.pubsub.PublishKafkaRecord_2_6$1.process(PublishKafkaRecord_2_6.java:521)
	at org.apache.nifi.controller.repository.StandardProcessSession.read(StandardProcessSession.java:2693)
	at org.apache.nifi.controller.repository.StandardProcessSession.read(StandardProcessSession.java:2661)
	at org.apache.nifi.processors.kafka.pubsub.PublishKafkaRecord_2_6.onTrigger(PublishKafkaRecord_2_6.java:513)
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27)
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1357)
	at org.apache.nifi.controller.tasks.ConnectableTask.invoke(ConnectableTask.java:246)
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:102)
	at org.apache.nifi.engine.FlowEngine$2.run(FlowEngine.java:110)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

One other thing that I noticed, though it's minor: When I change the "Publish Strategy" to "Use Wrapper" I'm given the option to configure the "Record Key Writer" property. But this is way down the list, about 8 or 10 properties down, so it's not at all obvious that it's made available. Probably want to move the "Record Key Writer" just after the "Publish Strategy" property sends it depends on it.

markap14 commented:
I also tried changing the "Publish Strategy" to "Use Record Content as Value" so that I could set the "Key Field." I then set the Key Field to `key`. And then changed the Publish Strategy back to "Use Wrapper".
This time, because the Key Field was specified, it sent the data. But what it sent was not correct.
It sent the key correctly. But for the Kafka message, it sent the entire payload, not just the `value` section.
And it sent no headers. So it appears to behave as if the Publish Strategy still was set to "Use Record Content as Value"

greyp9 commented:
> Thanks for updating @greyp9 . Trying this out again. I just tried sending the following content via PublishKafkaRecord_2_6:

> I set Publish Strategy to "Use Wrapper" and used a JsonTreeReader as the record reader. But I encountered a NullPointerException:
> 

Updated to validate content of message key.


> One other thing that I noticed, though it's minor: When I change the "Publish Strategy" to "Use Wrapper" I'm given the option to configure the "Record Key Writer" property. But this is way down the list, about 8 or 10 properties down, so it's not at all obvious that it's made available. Probably want to move the "Record Key Writer" just after the "Publish Strategy" property sends it depends on it.

Moved to just after "Publish Strategy".

greyp9 commented:
> ```
> {
>   "key": {
>     "type": "person"
>   },
>   "value": {
>     "name": "Mark",
>     "number": 49
>   },
>   "headers": {
>      "headerA": "headerAValue"
>   }
> }
> ```

Thanks.  There are enough permutations related to these changes such that I'm not sure what should be emitted for all of them.  I've made some simplifying assumptions; those may need adjustments.

Maybe a concrete example would help.  Given the record above, and the processor PublishKafkaRecord_2_6:

[Processor Property] PublishStrategy
```
if (Use Wrapper) {
    Kafka Record:
    Key = ?
    Value = ?
    Headers = ?
} else {
    Kafka Record:
    Key = ?
    Value = ?
    Headers = ?
}
```

---

[Processor Property] Attributes To Send as Headers (Regex)
```
if (Use Wrapper) {
    send attributes in wrapper record (do not add to kafka headers)
} else {
    send attributes in kafka headers
}
```
is this high-level logic valid?

---

[Processor Property] Message Key Field
```
if (Use Wrapper) {
    ignore (null key)
} else {
    send key (if record field exists) in Kafka key
}
```

is this right?

markap14 commented:
@greyp9 So, given the above (i will address in reverse order because I think you started with the hardest to describe and got easier as you went down the list) :) ...

The last point, about "Message Key Field" property I think is perfectly accurate.

As for the headers:
- I think what you said is accurate, but to clarify:
- If (Use Wrapper) - the headers to send would be a single header. Its name would be "headerA" and its value would be "headerAValue". FlowFile attributes would not be sent as headers.
- Else, the headers would be any FlowFile attribute that matches the "Attributes to Send as Headers (Regex)" property

Now, as for the other...
```
if (Use Wrapper) {
    Kafka Record:
    Key = { "type": "person" }
    Value = { "name": "Mark", "number": 49 }
    Headers = A single header with name "headerA", value "headerAValue"
} else {
    Kafka Record:
    Key = <Depends on the value of the 'Message Key Field' property>
    Value = <The entire JSON payload. I.e.,:>
    {
                       "key": {
                           "type": "person"
                        },
                       "value": {
                            "name": "Mark",
                            "number": 49
                       },
                       "headers": {
                             "headerA": "headerAValue"
                       }
     }
    Headers = <Whatever matches the 'Attributes to Send as Headers (Regex)' property>
}
```

So, in short, if Use Wrapper, the incoming FlowFile must have 3 fields:
Key. This becomes the kafka message key.
Value. This becomes the contents of the kafka message.
Headers. This becomes the headers attached to the kafka message.
Any other fields, such as metadata, would be ignored.

If NOT using wrapper, it would function as it always has. The entire contents of the Record go as the kafka message payload. The key and headers are determined based on the configured "Message Key Field" and "Attributes to Send as Headers (Regex)" properties.

Does that make sense? Or have I confused things even worse? :)

greyp9 commented:
> Does that make sense? Or have I confused things even worse? :)

Much clearer now; thanks.  I've added a new unit test that calls out the output differences between the publisher strategies VALUE and WRAPPER, and I leveraged this to adjust the wrapper to fit the expected behavior.  (And picked up some useful jackson knowledge along the way.)  :)

It should be easier to add test cases if we notice more discrepancies from the intended outputs.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/ConsumerLease.java:
```diff
@@ -591,51 +607,18 @@ private void writeRecordData(final ProcessSession session, final List<ConsumerRe
-                                handleParseFailure(consumerRecord, session, re, "Failed to write message from Kafka using the configured Record Writer. "
-                                    + "Will route message as its own FlowFile to the 'parse.failure' relationship");
-                                continue;
+                            if ((OutputStrategy.USE_WRAPPER.equals(outputStrategy)) && (consumerRecord.key() != null)) {
```
We don't need the `&& (consumerRecord.key() != null)` check here. We should be using the wrapper even if the key is null.

markap14 commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/ConsumerLease.java:
```diff
@@ -653,6 +636,131 @@ private void writeRecordData(final ProcessSession session, final List<ConsumerRe
+            }
+        } else if (KafkaProcessorUtils.KEY_AS_STRING.getValue().equals(keyFormat)) {
+            final RecordField recordField = new RecordField("key", RecordFieldType.STRING.getDataType());
+            tuple = new Tuple<>(recordField, new String(key, StandardCharsets.UTF_8));
```
If the original key was null, we should not be using `new String(key, StandardCharsets.UTF_8)` because that generates an empty string. Instead, we should just use `null`.

markap14 reviewed (commented):
Hey @greyp9 thanks for the updates. I did some testing. I need to do more testing. But the only issue that I ran into so far was that the wrapper is not being used if the key is null. A null key is perfectly valid, though, and the wrapper should still be used.

exceptionfactory commented on nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-2-6-processors/src/main/java/org/apache/nifi/processors/kafka/pubsub/PublisherLease.java:
```diff
@@ -188,18 +202,30 @@ void publish(final FlowFile flowFile, final RecordSet recordSet, final RecordSet
+                    headers = toHeadersWrapper(record.getValue("headers"));
+                    final Object key = record.getValue("key");
+                    final Object value = record.getValue("value");
+                    logger.trace("Key: {}", key);
+                    logger.trace("Value: {}", value);
```
Recommend removing these trace logs from the final version.
